### DISCUSSION TITLE
How to properly render a server side pagination in Storybook?

### DISCUSSION THREAD

**Discussion by SalahAdDin** [47ng/nuqs#539] [Q&A] [answered]
We are working with Storybook to visualize our components before using them on the pages.

We created our Pagination component using the example, specifically the server-side one, and we are trying to render it on Storybook, but it fails with the following error:
```bash
Error
[nuqs] Empty search params cache. Search params can't be accessed in Layouts.
See https://err.47ng.com/NUQS-500
Call Stack
 all
  navigation-Pagination-Pagination-stories.iframe.bundle.js:682:13
 Pagination
  navigation-Pagination-Pagination-stories.iframe.bundle.js:452:105
 renderWithHooks
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:125385:27
 mountIndeterminateComponent
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:131146:13
 beginWork$1
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:132746:16
 callCallback
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:134862:14
 invokeGuardedCallbackImpl
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:134911:16
 invokeGuardedCallback
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:134986:29
 beginWork
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:141177:28
 performUnitOfWork
  vendors-node_modules_pnpm_faker-js_faker_8_4_1_node_modules_faker-js_faker_dist_esm_index_mjs-c99bd3.iframe.bundle.js:140001:12
```

From the documentation, we understand we need to mock something, but, what is it?

Thank you!

**Comment by franky47** [comment, marked as answer]
According to the error URL, the issue is that the search params cache hasn't been populated. This can be done by calling the `parse` method on it, feeding it your mock search params, which will then be distributed to the RSC tree.

Documentation: https://nuqs.47ng.com/docs/server-side

**Reply by SalahAdDin** [reply]
Well, all the values on search params have a default value.

I wrapped the component with a layout, and it does not work yet.

**Reply by franky47** [reply]
Could you show me what your story file and relevant Storybook setup looks like please?

> Well, all the values on search params have a default value.

That's a valid point, the cache could be populated with default/null values obtained from the parsers (essentially parsing an empty search params object at initialisation), rather than throwing this error. The current behaviour is to use a frozen object to ensure the call to `parse` is the actual initialiser, and that those values can't ever be mutated by RSCs. So this sounds like a question of tradeoff between security and convenience.

> I wrapped the component with a layout, and it does not work yet.

The error is about Next.js layout files, not sure how Storybook handles those, but in Next.js RSCs won't re-render on search params change if they are children of a Layout. They need to be children of a Page and use the searchParams page prop to correctly re-render on URL changes.

https://nextjs.org/docs/app/api-reference/file-conventions/layout#layouts-do-not-receive-searchparams

**Reply by SalahAdDin** [reply]
We need only to show the component, we don't care about the navigation right now.

So, the Storybook goes like this:
```
import RootLayout from "@/app/[locale]/layout";
import { Meta, StoryObj } from "@storybook/react";

import { Pagination } from "./Pagination";

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  decorators: [
    (Story) => (
      <RootLayout
        params={{
          locale: "en",
        }}>
        <Story />
      </RootLayout>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Navigation: Story = {
  args: { pathname: "/search", page: 5, total: 7 },
};
```

**Reply by franky47** [reply]
Ok, so using the playground code as a reference, a fix could look like this (assuming the decorators run as server components by default).

```tsx
import RootLayout from "@/app/[locale]/layout";
import { Meta, StoryObj } from "@storybook/react";

import { Pagination } from "./Pagination";
import { searchParamsCache } from './searchParams'

function SearchParamsCacheProvider({ children }: { children: any }) {
  searchParamsCache.parse({})
  return children
}

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  decorators: [
    (Story) => (
      <RootLayout
        params={{
          locale: "en",
        }}>
        <SearchParamsCacheProvider>
          <Story />
        </SearchParamsCacheProvider>
      </RootLayout>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Navigation: Story = {
  args: { pathname: "/search", page: 5, total: 7 },
};
```

**Reply by SalahAdDin** [reply]
It worked pretty well, thanks!

Is it possible to use it without a `layout`?

**Reply by franky47** [reply]
> Is it possible to use it without a `layout`?

If by layout you mean a Next.js layout component, yes, as mentioned in the error searchParams are actually not supported in layouts, only in pages.

**Reply by SalahAdDin** [reply]
Actually, It is not required, that's good.

BUt if I go to another story that uses the search params cache provider, it breaks: 
![image](https://github.com/47ng/nuqs/assets/5034215/2238eab0-6a75-45bd-8103-5f89b24cd0ec)

**Reply by SalahAdDin** [reply]
@franky47 Is it an issue with `nuqs` or with the provider itself?
